77.md

```diff
@@ -0,0 +1,175 @@
+
+Since Negentropy is a binary protocol, this wrapper hex-encodes its messages. The specification for Negentropy Protocol V1 is attached as an appendix to this NIP below.
+
+## High-Level Protocol Description
+
+We're going to call the two sides engaged in the sync the client and the relay (even though the initiator could be another relay instead of a client).
+
+* (1) Client (initiator) chooses a filter, and retrieves the set of events that it has locally that match this filter (or uses a cache), and constructs an initial message.
+* (2) Client sends a `NEG-OPEN` message to the relay, which includes the filter and the initial message.
+* (3) Relay selects the set of events that it has locally that match the filter (or uses a cache).
+* (4) Relay constructs a response and returns it to the client in a `NEG-MSG` message.
+* (5) Client parses the message to learn about IDs it has (and relay needs) and IDs it needs (and relay has).
+  * If client wishes to continue, then it constructs a new message and sends it to the relay in a `NEG-MSG` message. Goto step 4.
+  * If client wishes to stop, then it sends a `NEG-CLOSE` message or disconnects the websocket.
+
+The above protocol only results in the client learning about IDs it has/needs, and does not actually transfer events. Given these IDs, the client can upload events it has with `EVENT`, and/or download events it needs with `REQ`. This can be performed over the same websocket connection in parallel with subsequent `NEG-MSG` messages. If a client is only interested in determining the number of unique events (ie, reaction counts), it may choose to not download/upload at all.
+
+## Nostr Messages
+
+### Initial message (client to relay):
+
+```jsonc
+[
+    "NEG-OPEN",
+    <subscription ID string>,
+    <filter>,
+    <initialMessage, hex-encoded>
+]
+```
+
+* The subscription ID is used by each side to identify which query a message refers to. It only needs to be long enough to distinguish it from any other concurrent subscriptions on this websocket connection (an integer that increments once per `NEG-OPEN` is fine). Subscription IDs are in a separate namespace from `REQ` subscription IDs. If a `NEG-OPEN` is issued for a currently open subscription ID, the existing subscription is first closed.
+* The filter is as described in [NIP-01](01.md).
+* `initialMessage` is the initial Negentropy binary message, hex-encoded. See appendix.
+
+### Error message (relay to client):
+
+If a request cannot be serviced by the relay, an error is returned to the client:
+
+```jsonc
+[
+    "NEG-ERR",
+    <subscription ID string>,
+    <reason code string>
+]
+```
+
+Error reasons are the same format as in NIP-01. They should begin with a machine-readable single-word prefix, followed by a `:` and then a human-readable message with more information.
+
+The current suggested error reasons are
```

Comment:
Should we also add something like `invalid: unknown subscription id`?  AFAIK there is no equivalent in NIP-01, since it's not needed. In case of `CLOSE` message the client is not expecting any reply, so it's fine the relay doesn't send back an error, but in case of `MSG` the client is actually waiting for a response, so it makes send to send appropriate error.

Reply:
Good question. Currently strfry sends `closed: unknown subscription handle` in this case. It doesn't bother trying to differentiate between a subscription ID that was previously open but got closed (by a `NEG-CLOSE` message, timeout, etc) and one that was never opened in the first place.